Custom Salt States (#807)
uyuni-project/uyuni-docs: modules/salt/pages/custom-states.adoc

[[custom-states]]
= Custom Salt States

You can create your own custom Salt states with {productname} as centrally managed configuration channels.

> Custom states are stored as Salt state files on the {productname} Server with a ``.sls`` extension, in the [path]``/srv/susemanager/salt/`` directory.

Reviewer: `/srv/susemanager/salt/` has content which is managed by SUMA. I think no customer should mess around with files in this directory. So I am not sure if it is wise to tell them where it is. At least not at such a prominent place?

== Create a New Custom Salt Channel

You can use the {productname} {webui} to create and edit custom Salt state files.

> You must create a state channel first, with an initial state named ``init.sls``.

Reviewer 2: As a complete noob, I understand this as 
```
you must create the state channel first, and then you need to
create the `init.sls` file
```
but apparently this states file will be created automatically when you create the state channel.

If that is so, I would rephrase last sentence as.
```
You must create a state channel first. This newly created channel
will contain an initial states file named ``init.sls``.
```

@juliogonzalez can you please confirm I'm not saying crap, here?

Author: When I went through the process, it's kind of not automatically created, though. I need to type something in the box.

Reviewer 2: thanks for the explanation Lana

The ``init.sls`` file is used to reference all other state files within the channel.
The custom states that you create using the {webui} are stored on the {productname} Server in the the [path]``/srv/susemanager/salt/<organization>/`` directory.

After the channel is created with an ``init.sls`` file, you can write additional state files in the {webui}.
Alternatively, you can upload existing state files to use within your state channel, or import them from other channels or clients.

.Procedure: Creating a Custom Salt Channel and Initial State
. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click btn:[Create State Channel].
. In the [guimenu]``Name`` field, type a name for your state.
. In the [guimenu]``Label`` field, type a label.
  Use alphanumeric characters, hyphens, and underscores.
  Do not use spaces.
. In the [guimenu]``Description`` field, type a short description of the configuration your state performs.
. In the [guimenu]``SLS Contents`` field, type the contents of your ``init.sls`` state.
  Ensure your file starts by specifying the source of the managed file, using this syntax:
+
----
file.managed:

>   - source: salt://<org_name>/<channel_name>/etc/<filename>

Reviewer: There is always an "ID" required and in most of the cases also a "name".

. To copy an existing file, click btn:``Import a File from Another Channel or System`` and select the file to copy.

.Procedure: Editing a Custom Salt State
. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click btn:[View/Edit <filename>.sls File].
. Make your changes to the file
. Click btn:[Update Configuration File] to save your state.

You can also manage revisions, compare the state to others in your organization, and download the ``.sls`` file from this dialog.

.Procedure: Assigning a Client to a Custom Salt State
. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click the name of the state you want to assign a client to.
. Navigate to the menu:Systems[Target Systems] tab.
. Check the clients you want to assign.
. Click btn:[Subscribe systems].

== Example Custom State Files

This section contains some example custom state files.
Use these as a basis for writing your own custom states.

.Example: Initial State File

----
file.managed:
  - source: salt://example_org/example_channel/etc/motd
[
  FIXME: reference other file here somehow

> ]

Reviewer: ID and name is missing

----

.Example: Custom State File

----
install required packages:
  pkg.installed:
    - pkgs:
      - lsof
    - failhard: True
update packages:
  module.run:
    - name: pkg.upgrade
    - failhard: True
check if reboot is needed:
  salt.set_reboot_needed:
    - failhard: True
reboot:
   salt.reboot_if_needed:

>      - failhard: True

Reviewer: where does this example come from? Is it how we want our customers to do things?

Author: One of the Aussie devs provided this to me when I couldn't find one, and I edited the hell out of it. I have no idea what we want to tell customers, this is the best I could do with what I had to work with. Any/all suggestions welcome!

Reviewer: Please remove this example. At least the last states `salt.set_reboot_needed` and `salt.reboot_if_needed` seems to be unavailable. They do not exist.

A state keeping things up to date is nice, but hard to write. We better do not describe something which is not working.

----

== Custom State to Trust a GPG Key

By default, operating systems trust only their own GPG keys when they are installed, and do not trust keys provided by third party packages.
The clients can be successfully bootstrapped without the GPG key being trusted.

> However, you cannot install new client tool packages or update them until the keys are trusted.

Reviewer: ```suggestion
However, you cannot install new third party packages or update them until the keys are trusted.
```

Salt clients are set to trust {suse} tools channels GPG keys when they are bootstrapped.

> For all other clients and channels, you need to manually trust third party GPG keys.

Reviewer: I would suggest from now on we talk about 3rd party channels and packages. We better use not examples with SLES or SUSE Manager keys prevent the assumption that other SUSE provided keys must be trusted on the clients.

If you are bootstrapping Salt clients from the {productname} {webui}, you can use a custom Salt state to trust the GPG key.

.Procedure: Trusting a GPG Key With a Custom Salt State
. On the {productname} Server, at the command prompt, check the contents of the [path]``/srv/www/htdocs/pub/`` directory.
  This directory contains all available public keys.

>   Take a note of the key that applies to the channel assigned to the client you are registering.

Reviewer: This directory contains keys we ship, but in most cases these are not the keys the customer has to trust manually.
We should concentrate on cases where a complete new key is needed.

1. find the key which needs to be trusted. Check the web page of the vendor or ask a key server when the keyid is known. Check the fingerprint to verify, that the correct key is used.
2. store the key at a place where client can download it. For example `/srv/www/htdocs/pub/` where we also store our keys

... continued below

. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click btn:[Create State Channel].
. In the [guimenu]``Name`` field, type a name for your state.
  For example, ``GPG Key Trusts``
. In the [guimenu]``Label`` field, type a label.
  For example, ``GPG_Key_Trusts``
. In the [guimenu]``Description`` field, type a short description of the configuration your state performs.
  For example, ``Trusts GPG Keys for CentOS``.
. In the [guimenu]``SLS Contents`` field, create a state to retrieve the appropriate key from the {productname} Server and trust it on the client.
  For example FIXME:
+
----
cmd.run:
  - salt '*' gpg.import_key filename='/srv/www/htdocs/pub/centos8-gpg-pubkey-05B555B38483C65D.key'

>   - salt '*' gpg.trust_key keyid='centos8-gpg-pubkey-05B555B38483C65D.key' trust_level='trusted'

Reviewer 2: Using Salt to call Salt seems a bit wird to me... Isn't there something more elegant and straightforward?

Author: This was a wild guess on my part. I need someone who knows what they're doing here :stuck_out_tongue:

Author: Perhaps @juliogonzalez can help?

Author: @juliogonzalez if you can't help, or don't have bandwidth, can you recommend someone else who could?

Reviewer: I would need to check, but i fear this would not do what you expect.

Author: No, I wasn't really expecting that it would. Who could tell me?

Reviewer: Problem: It would be better if we would have a salt module for trusting GPG keys on different OSes, but this does not exist yet. So trusting GPG keys depends on the OS!

```
rpm_trust_gpg_key:
  cmd.run:
    - name: rpm --import https://{{ salt['pillar.get']('mgr_server') }}/pub/third-party-gpg.key
    - unless: rpm -q gpg-pubkey-<key_id>

deb_trust_gpg_key:
  mgrcompat.module_run:
    - name: pkg.add_repo_key
    - path: https://{{ salt['pillar.get']('mgr_server') }}/pub/third-party-gpg.key
```

Reviewer: Of cause you can also add the GPG key files to the config channel, use a `file.managed` to deploy them on the minion and use a local path instead of a `https://` URL .

. To create a new file, click btn:``Create configuration file`` and type the contents of the file.
. To upload an existing file, click btn:``Upload Configuration Files`` and select the file to upload.
. To copy an existing file, click btn:``Import a File from Another Channel or System`` and select the file to copy.

.Procedure: Editing a Custom Salt State
. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click btn:[View/Edit <filename>.sls File].
. Make your changes to the file
. Click btn:[Update Configuration File] to save your state.

You can also manage revisions, compare the state to others in your organization, and download the ``.sls`` file from this dialog.

.Procedure: Assigning a Client to a Custom Salt State
. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click the name of the state you want to assign a client to.
. Navigate to the menu:Systems[Target Systems] tab.
. Check the clients you want to assign.
. Click btn:[Subscribe systems].

== Example Custom State Files

This section contains some example custom state files.
Use these as a basis for writing your own custom states.

> .Example: Initial State File

Reviewer: ```suggestion
.Example: Manage a File
```

Initial State File is just a state with the name of the channel. It has no special content.
It can contain whatever the customer want to manage. There is no need to manage additional files at all.

I would suggest to list here a number of the mostly used state modules.

- file.managed
- pkg.installed
- cmd.run
- service.running

> ----
> file.managed:
>   - source: salt://example_org/example_channel/etc/ID/filename
> [
>   FIXME: reference other file here somehow
> ]
> ----

Reviewer: ```suggestion
----
my_config_change_id:
  file.managed:
    -name: /etc/my.conf
    - source: salt://example_org/example_channel/etc/my.conf
    - user: root
    - group: root
    - mode: 644
    - template: jinja
----
```

> 

Reviewer: ```suggestion
.Example: Package Management

----
my_pkg_id:
  pkg.installed:
    - refresh: True
    - pkgs:
      - glibc
      - kernel-default
      - hello: 1.0-42
----

.Example: Remote Command

----
ip_forward-on:
  cmd.run:
    - name: echo "1" > /proc/sys/net/ipv4/ip_forward
    - onlyif:
      - test `cat /proc/sys/net/ipv4/ip_forward` -eq 0
----

.Example: Service Management

----
time_service_id:
  service.running:
    - name: chronyd
    - enable: True
----
```

. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click btn:[Create State Channel].
. In the [guimenu]``Name`` field, type a name for your state.
. In the [guimenu]``Label`` field, type a label.
  Use alphanumeric characters, hyphens, and underscores.
  Do not use spaces.
. In the [guimenu]``Description`` field, type a short description of the configuration your state performs.
. In the [guimenu]``SLS Contents`` field, type the contents of your ``init.sls`` state.
  If you want to reference file templates in this configuration channel, ensure your file starts by specifying the source of the managed file, using this syntax:
+
----
file.managed:
  - source: salt://<org_name>/<channel_name>/etc/<ID>/<filename>
----
+
 Example custom state files are given later in this section.
. Click btn:[Update Channel] to save your state.

.Procedure: Adding Additional Files to a Custom State Channel
. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click the name of the channel you want to add files to.
. To create a new file, click btn:``Create configuration file`` and type the contents of the file.

> . To upload an existing file, click btn:``Upload Configuration Files`` and select the file to upload.

Reviewer 3: ```suggestion
. To upload an existing file, click btn:[Upload Configuration Files] and select the file to upload.
```

> . To copy an existing file, click btn:``Import a File from Another Channel or System`` and select the file to copy.

Reviewer 3: ```suggestion
. To copy an existing file, click btn:[Import a File from Another Channel or System] and select the file to copy.
```

.Procedure: Editing a Custom Salt State
. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click btn:[View/Edit <filename>.sls File].

> . Make your changes to the file

Reviewer 3: ```suggestion
. Make your changes to the file.
```

. Click btn:[Update Configuration File] to save your state.

You can also manage revisions, compare the state to others in your organization, and download the ``.sls`` file from this dialog.

.Procedure: Assigning a Client to a Custom Salt State
. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click the name of the state you want to assign a client to.
. Navigate to the menu:Systems[Target Systems] tab.
. Check the clients you want to assign.
. Click btn:[Subscribe systems].

> For more information about Salt state modules, see https://docs.saltproject.io/en/latest/ref/states/all/index.html[].

Reviewer 3: ```suggestion
For more information about Salt state modules, see https://docs.saltproject.io/en/latest/ref/states/all/index.html.
```

== Example Custom State Files

This section contains some example custom state files.
Use these as a basis for writing your own custom states.

.Example: Manage a File

----
my_config_change_id:
  file.managed:

>     -name: /etc/my.conf

Reviewer 3: ```suggestion
    - name: /etc/my.conf
```

The clients can be successfully bootstrapped without the GPG key being trusted.
However, you cannot install new third party packages or update them until the keys are trusted.

Salt clients are set to trust {suse} tools channels GPG keys when they are bootstrapped.
For all other clients and channels, you need to manually trust third party GPG keys.

If you are bootstrapping Salt clients from the {productname} {webui}, you can use a custom Salt state to trust the GPG key.

.Procedure: Trusting a GPG Key With a Custom Salt State
. Locate the key that you need to trust.
  Ensure you have the correct key, and that you also have the fingerprint used to verify the key.
  This information is available from the vendor or, in some cases, from a key server.
. Copy the key to a file location where the client can access it.
  We recommend saving it in the [path]``/srv/www/htdocs/pub/`` directory, where all {suse} public keys are also saved.
. In the {productname} {webui}, navigate to menu:Configuration[Channels].
. Click btn:[Create State Channel].
. In the [guimenu]``Name`` field, type a name for your state.

>   For example, ``GPG Key Trusts``

Reviewer 3: ```suggestion
  For example, ``GPG Key Trusts``.
```

. In the [guimenu]``Label`` field, type a label.

>   For example, ``GPG_Key_Trusts``

Reviewer 3: ```suggestion
  For example, ``GPG_Key_Trusts``.
```
